doc/design_docs/packed_conditional_fields.md

```diff
@@ -0,0 +1,205 @@
+# Design Sketch: Packed Conditional Fields
+
+## Motivation
+
+The [`$next` keyword](archive/next_keyword.md) was introduced to facilitate
+writing Emboss definitions for packed or mostly-packed structs, while avoiding
+
+ - the potential error of unintentionally introducing a gap or overlap by
+   using the wrong start location,
+ - complexity of computing the start location of variable sized fields, and
+ - including "redundant" and potentially hardcoded values that must be updated
+   separately if the dependent values are updated.
+
+However, one use case not covered by the existing `$next` keyword is that it
+does not keep a structure tightly packed if there are conditional fields. For
+instance, the structure in
+[Example 1](#example-1-conditional-structure-with-next) will have a 4 byte gap
+if `type_code` is `1`, even though the `metadata` field is not present.
+
+This tends to be non-intuitive, the author of this design sketch has seen
+multiple authors of emboss definitions write `$next` after conditional fields
+in this manner and it always results in surprise when it does not work how they
+intended, which in this example is to tightly pack `payload` to follow after
+`type_code` when `metadata` is not present.
+
+###### Example 1. Conditional structure with $next
+```
+struct ConditionalExample:
+  0       [+2]  UInt       type_code
+  if type_code == 0x01:
+    $next [+4]  UInt:8[4]  metadata
+  $next   [+10] Uint:8[4]  payload
+```
+
+## Considerations
+
+There are a few considerations for implementing the behavior that developers
+intend when they write code like 
+[Example 1](#example-1-conditional-structure-with-next) that are covered in
+this section.
+
+### Backwards Compatibility
```

Comment:
I have two major thoughts on this:

1.  There probably are not any (correct) uses of `$next` after a conditional field in the wild (I did not turn up any in an internal audit).
2.  In principle, a breaking change should break *noisily*.  In this case, that means turning any use of `$next` immediately after a conditional field into an error, and either using a new keyword (`$pack` is fine) or having users opt into the new behavior (`[next_keyword = "Pack"]`).  Compilation errors are annoying, but not nearly as problematic as silently changing behavior.